### PR TITLE
FHIR Bulk Export script

### PR DESCRIPTION
**Pull request by ivan-c**
- FHIR Bulk Export script
  - Pass FHIR base URL as first argument
  - Downloads files to current directory (or given via argument)

#### TODO
- [x] Allow passing config vars (auth token, timeouts/paths) via argparse
- [x] Add references to [Bulk Export IG](https://build.fhir.org/ig/HL7/bulk-data/export.html)

### CODE REVIEW

File: utils/fhir-server-export.py [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+        if not retry_after:
+            return status_poll_response.json()
+
+        progress = status_poll_response.headers.get("X-Progress")
+        if progress:
+            print("progress: ", progress)
+
+        print(f"waiting {retry_after} seconds")
+        rety_time += retry_after
+        time.sleep(retry_after)
```
Comment (daniellrgn) [inline utils/fhir-server-export.py:43]:
```suggestion
    max_retry_time = 10 * 60
    retry_time = 0

    while retry_time < max_retry_time:
        status_poll_response = requests.get(status_poll_url)
        status_poll_response.raise_for_status()

        retry_after = int(status_poll_response.headers.get("Retry-After", 0))
        if not retry_after:
            return status_poll_response.json()

        progress = status_poll_response.headers.get("X-Progress")
        if progress:
            print("progress: ", progress)

        print(f"waiting {retry_after} seconds")
        retry_time += retry_after
        time.sleep(retry_after)
```

**Review by daniellrgn — APPROVED** [review]
Looks good to me!

**Comment by mcjustin** [comment]
@ivan-c  I tried this per https://cirg.slack.com/archives/C01P9V67NMA/p1691255466471439?thread_ts=1691083626.198519&cid=C01P9V67NMA
... and got:

```
debadmin@DESKTOP-0I2L7PU:~/fhir-eval-environments/utils$ python3 fhir-server-export.py https://fhir.isacc.dev.cirg.uw.edu/fhir
progress:  Build in progress - Status set to SUBMITTED at 2023-08-16T21:19:26.806+00:00
waiting 120 seconds
Traceback (most recent call last):
  File "fhir-server-export.py", line 111, in <module>
    main()
  File "fhir-server-export.py", line 92, in main
    for file_item in complete_json.get("output"):
TypeError: 'NoneType' object is not iterable
```

Any ideas?

File: utils/fhir-server-export.py [hunk 2]
```diff
@@ -0,0 +1,111 @@
...
+def fixup_url(url, base_url):
+    """
+    Replace FHIR base URL in given FHIR API call with different base_url
+    Helpful when a FHIR server is configured with a server_name that does not match its public one
+    """
+    if not url.startswith(base_url):
+        second_last_path, last_path = url.split("/")[-2:]
+        # FHIR operations
+        if last_path.startswith("$"):
+            return f"{base_url}/{last_path}"
```
Comment (pbugni) [inline utils/fhir-server-export.py:31]: i don't know the spec as well as you, but is this a safe assumption?  do operations never have a named resource in the request, i.e. "http://base/fhir/Patient/$process"

Comment (ivan-c) [reply utils/fhir-server-export.py:36]: it's not generally a safe assumption. however, I haven't been able to figure out a consistent way to get the FHIR base URL, so I've been relying on some loose heuristics, specific to bulk export.

would love to hear any ideas, but will update the docstring to make it more obvious this process isn't perfect

File: utils/fhir-server-export.py [hunk 3]
```diff
@@ -0,0 +1,111 @@
...
+                f.write(chunk)
+    return filename
+
+
+def fixup_url(url, base_url):
+    """
+    Replace FHIR base URL in given FHIR API call with different base_url
+    Helpful when a FHIR server is configured with a server_name that does not match its public one
+    """
+    if not url.startswith(base_url):
```
[the remark this answers is not in the document]
Comment (ivan-c) [reply utils/fhir-server-export.py]: thanks, done!

File: utils/fhir-server-export.py [hunk 4]
```diff
@@ -0,0 +1,111 @@
...
+        if progress:
+            print("progress: ", progress)
+
+        print(f"waiting {retry_after} seconds")
+        rety_time += retry_after
+        time.sleep(retry_after)
+
+
+def main():
+    # TODO replace with argparse
```
[the remark this answers is not in the document]
Comment (ivan-c) [reply utils/fhir-server-export.py:71]: would probably use click if this script wasn't so freestanding. was hoping to use this script in the DAWG environment, and I'm not very familiar with windows deploys, so trying to keep deployment stupid simple

File: utils/fhir-server-export.py (hunk 1 above)
Comment (ivan-c) [reply utils/fhir-server-export.py:67]: are there changes in this diff? I keep squinting at it, but nothing's standing out to me

Comment (mcjustin) [reply utils/fhir-server-export.py:67]: Not that I can tell, but I think your question was for @daniellrgn

**Review by mcjustin — APPROVED** [review]

**Comment by mcjustin** [comment]
@ivan-c  I don't recall whether we had a plan for deltas here... Can I tell it to only export resources which have changed since some time?

**Comment by mcjustin** [comment]
> @ivan-c I don't recall whether we had a plan for deltas here... Can I tell it to only export resources which have changed since some time?

I'm likely to add a `_since` parameter here, following the pattern you established with `_type`.

**Comment by mcjustin** [comment]
@ivan-c  I had a good run from cesium, the only problem was that it seems to have ignored the `type` param. 

Any ideas?

```
mcjustin@cesium fhir-eval-environments/utils$ ./fhir-server-export.py https://fhir-auth.uwmedicine.stage.cosri.app/fhir/ --directory . --auth-token $TOKEN --max-timeout 1200 --type QuestionnaireResponse
```

**Comment by ivan-c** [comment]
> @ivan-c I had a good run from cesium, the only problem was that it seems to have ignored the `type` param.
> 
> Any ideas?

@mcjustin 


Sorry, I *just* fixed that in 5fe9bed

**Comment by mcjustin** [comment]
> > @ivan-c I had a good run from cesium, the only problem was that it seems to have ignored the `type` param.
> > Any ideas?
> 
> @mcjustin
> 
> Sorry, I _just_ fixed that in [5fe9bed](https://github.com/uwcirg/fhir-eval-environments/commit/5fe9bed158921ec16c63b43b9263c90ecbf41a96)

@ivan-c  I actually did have that version already, pulled again just now to confirm and tried it again. I also added `since` but that's ignored too (works like a charm otherwise). Did `type` work for you?

**Comment by ivan-c** [comment]
> I actually did have that version already, pulled again just now to confirm and tried it again. I also added `since` but that's ignored too (works like a charm otherwise). Did `type` work for you?

@mcjustin 

I hadn't tested it, but after adding some debugging, it looks like `_type` is being ignored by Hapi. Not clear why though...

**Comment by mcjustin** [comment]
> I hadn't tested it, but after adding some debugging, it looks like `_type` is being ignored by Hapi. Not clear why though...

@ivan-c  I wonder if it should be `_typeFilter` - per  https://github.com/hapifhir/hapi-fhir/pull/2797 ,  though https://hapifhir.io/hapi-fhir/docs/introduction/changelog.html has mixed messaging... I'm going to give `_typeFilter` a shot

**Comment by ivan-c** [comment]
>I wonder if it should be `_typeFilter` - per [hapifhir/hapi-fhir#2797](https://github.com/hapifhir/hapi-fhir/pull/2797) , though https://hapifhir.io/hapi-fhir/docs/introduction/changelog.html has mixed messaging... I'm going to give `_typeFilter` a shot

@mcjustin 
Couldn't hurt to try!

I think `_typeFilter` is for filtering individual FHIR resources based on some criteria, but if it works, we might be able to filter on FHIR `resourceType`

>a server with support for the parameter and requested search queries SHALL filter the data in the response to only include resources that meet the specified criteria

https://build.fhir.org/ig/HL7/bulk-data/export.html#query-parameters

**Comment by ivan-c** [comment]
I see references to the required (non-optional) querystring parameters in the [Conformance statement](https://fhir-auth.uwmedicine.stage.cosri.app/fhir/metadata); surprising that they don't work :disappointed: 

```
    {
      "name": "_type",
      "use": "in",
      "min": 0,
      "max": "1",
      "type": "string"
    },
```
https://fhir-auth.uwmedicine.stage.cosri.app/fhir/OperationDefinition/GroupPatient-its-export

**Comment by mcjustin** [comment]
@ivan-c  are you able to upgrade Hapi from 6.2.2 to 6.8.2 by Wed mid-day?

This fix was in 6.4.1 and I see other bulk export related changes later:
[#4891](https://github.com/hapifhir/hapi-fhir/issues/4891) | *Initiating a bulk export with a _type filter would  sometimes return resource types not specified in the filter. This has  been fixed.*

File: utils/fhir-server-export.py [hunk 5]
```diff
@@ -0,0 +1,151 @@
...
+        print("server-side caching disabled")
+        headers["Cache-control"] = "no-cache"
+
+    if auth_token is not None:
+        headers["Authorization"] = f"Bearer {auth_token}"
+
+    if type is not None:
+        params["_type"] = type
+
+    kickoff_response = requests.post(
```
Comment (ivan-c) [inline utils/fhir-server-export.py:95]: @mcjustin finally figured out this was why Hapi was ignoring our [parameters](https://build.fhir.org/ig/HL7/bulk-data/export.html#query-parameters): I misread the Bulk Export spec. 

support for kickoff requests via `GET` is required. `POST` kickoff requests are also allowed, but use a different format (POSTing FHIR Parameters JSON data)

>A server SHALL support GET requests and MAY support POST requests that supply parameters using the FHIR [Parameters Resource](https://www.hl7.org/fhir/parameters.html)

https://build.fhir.org/ig/HL7/bulk-data/export.html#bulk-data-kick-off-request

Doing some more testing but should have a fix soon

**Comment by mcjustin** [comment]
@ivan-c  I tried this with and without `_type` and `_since` params but it's not giving me any output.

```
mcjustin@cesium fhir-eval-environments/utils$ ./fhir-server-export.py https://fhir-auth.uwmedicine.stage.cosri.app/fhir/ --auth-token $TOKEN --max-timeout 1200 --type QuestionnaireResponse --since  2023-10-17T05:00:00Z

error: export completed successfully, but response is not JSON:


mcjustin@cesium fhir-eval-environments/utils$ ./fhir-server-export.py https://fhir-auth.uwmedicine.stage.cosri.app/fhir/ --auth-token $TOKEN --max-timeout 1200

progress:  Build in progress - Status set to IN_PROGRESS at 2023-10-18T18:31:30.165+00:00
waiting 120 seconds
error: export completed successfully, but response is not JSON:
```

**Comment by ivan-c** [comment]
> I tried this with and without `_type` and `_since` params but it's not giving me any output.

@mcjustin 
I saw the same error when Hapi didn't have any data to export (eg _since in the future). 

Did the UWMC stage server get repopulated with paintracker data after I upgraded it?

Hapi should still return info via JSON in the situation, but I observed a completely empty response. Add a note in the error message in ff0ecbe

**Comment by mcjustin** [comment]
> > I tried this with and without `_type` and `_since` params but it's not giving me any output.
> 
> @mcjustin I saw the same error when Hapi didn't have any data to export (eg _since in the future).
> 
> Did the UWMC stage server get repopulated with paintracker data after I upgraded it?
> 
> Hapi should still return info via JSON in the situation, but I observed a completely empty response. Add a note in the error message in [81309f2](https://github.com/uwcirg/fhir-eval-environments/commit/81309f25ed8b7382355161876832363990e95d55)

@ivan-c  ah I see that the upgrade today wiped Hapi - I didn't think it would. Easy change to proceed here: I'll add Marcus and his QR's from PainTracker staging and try again.

**Comment by mcjustin** [comment]
Noting that this passes my testing today, including `--type`, `--destination`, and `--since` params as run from cesium (see above).

**Pull request merged**